Account balance management

amustapha opened the pull request:
- Allows setting the initial balance for backtester to begin
- Implements deposit/withdraw functions to 
- implement fetch balance ccxt method

amustapha committed:
implement account functionality

amustapha committed:
fix return value to comform with ccxt standard

amustapha committed:
make fee optional and fix empty dataframe warning

amustapha committed:
implement tests for backtester balances

amustapha committed:
make balances protected and init_balances private

amustapha committed:
setup flake8 and black

amustapha committed:
add pre-commit hooks for flake8, black and tests

amustapha committed:
update readme

amustapha committed:
add blank line at end of file

peterolayinka commented:
Looking into this.

peterolayinka commented on ccxt_backtesting_exchange/backtester.py:
```diff
@@ -0,0 +1,89 @@
+        self._update_asset_balance(asset, "free", amount)
+        self._update_asset_balance(asset, "total", amount)
+
+    def withdraw(self, asset: str, amount: float, id=None):
```
Beautiful PR, but I think we should try to use similar argument name as CCTX exchange class method for withdraw, in this case, I believe we should maintain the same arugument to prevent change of code when people switch between backtest ccxt and real excahnge ccxt, in our case asset should be `code`

<img width="220" alt="Screenshot 2025-02-02 at 16 28 43" src="https://github.com/user-attachments/assets/aa94fa2a-d14b-439f-ad1c-ed811ecdf0a7" />


> :param str code: unified currency code

peterolayinka commented on ccxt_backtesting_exchange/backtester.py:
```diff
@@ -0,0 +1,89 @@
+        self._update_asset_balance(asset, "free", -amount)
+        self._update_asset_balance(asset, "total", -amount)
+
+    def fetch_balance(self):
```
I also think we should add params to methods to enable compatibility, 

`params={}`

<img width="385" alt="Screenshot 2025-02-02 at 16 33 34" src="https://github.com/user-attachments/assets/835825ce-10ca-4d95-a8fc-47594581f0a3" />

Same for withdraw.

peterolayinka commented:
Based on the code, the ccxt backtest exhange now has the fetch_balance and withdraw method, with an added `deposit` method which is not in the core ccxt methods, but this seems to be needed in our case because we need a way to add money to the backtest outside initialisation since we can't deposit onchain.

peterolayinka committed:
Add params for compatibility sake.

This was if a code logic passed down params we don't have to modify the code, it just accepts it even though it does nothing, this way we prevent augument errors or we could use **kwargs, but I believe it is better to use same parameters used in the ccxt Excahnge itself.

peterolayinka committed:
'id' is not a parameter in the ccxt Exchange.

I could have left it for deposit, but I guess we aren't really using id here.

peterolayinka committed:
Use the same naming convention in ccxt Exchange

For public class method, I believe it is ideal to use the same parameter name, because some usage might be by passing the parameters as a named argument, and in this case it will fail or require a code change.

peterolayinka reviewed (approved):
Great work here, I believe this is good to go. I love how you handle state with pandas. I guess I have lot to learn about pandas.

amustapha commented:
Thanks for pointing out the issues and also fixing them. Na you try pass
